packages/flutter/lib/src/widgets/binding.dart

```diff
@@ -566,6 +594,57 @@ mixin WidgetsBinding on BindingBase, ServicesBinding, SchedulerBinding, GestureB
     return Future<void>.value();
   }
 
+  final Map<DebugPreHotRestartCallback, String> _hotRestartCallbacks = <DebugPreHotRestartCallback, String>{};
+
+  /// Register a callback that will be invoked before a hot restart is called.
+  ///
+  /// In non-debug modes this method is a no-op. This can be used to release native
+  /// resources acquired through platform channels or `dart:ffi`. Future returning
+  /// callbacks will be awaited, allowing for async tear downs.
+  ///
+  /// The following sample code shows how to use debugRegisterHotRestartCallback to handle
+  /// tearing down a native resource acquired through `dart:ffi`. In this example, if
+  /// the `context` pointer is not passed through to the `_destroyContext` function before
+  /// a hot restart, the application will crash after a hot restart.
+  ///
+  /// ```dart
+  /// import 'dart:ffi';
+  /// import 'package:flutter/foundation.dart';
+  /// import 'package:flutter/widgets.dart';
+  ///
+  /// final DynamicLibrary _lib = DynamicLibrary.open('some_native_lib.dll');
+  ///
+  /// final Pointer<NativeType> Function() _createContext = _lib.lookupFunction<
+  ///     Pointer<NativeType> Function(),
+  ///     Pointer<NativeType> Function()>('Native_create');
+  ///
+  /// final void Function(Pointer<NativeType>) _destroyContext = _lib.lookupFunction<
+  ///     Void Function(Pointer<NativeType>),
+  ///     void Function(Pointer<NativeType>)>('Native_destroy');
+  ///
+  /// class NativeResourceService {
+  ///   NativeResourceService() {
+  ///     if (kDebugMode) {
+  ///       WidgetsBinding.instance.debugRegisterHotRestartCallback(
+  ///           () => _destroyContext(_context),
+  ///           debugLabel: 'NativeResourceService');
+  ///     }
+  ///   }
+  ///
+  ///   /// Acquire native resources that must be released before they can
+  ///   /// be re-acquired.
+  ///   late final Pointer<NativeType> _context = _createContext();
+  /// }
+  /// ```
+  void debugRegisterHotRestartCallback(DebugPreHotRestartCallback callback, {String debugLabel = 'unknown'}) {
+    assert(kDebugMode, '''debugRegisterHotRestartCallback can only be called in debug mode. '''
+                       '''Use kDebugMode or wrap the call in an assert.''');
```

Comment:
no need for to use `'''` for one-line string literals

Reply:
in general though this assert isn't going to do anything -- asserts are removed when kDebugMode is false. :-)

I would recommend `if (!kDebugMode) { throw StateError(...); }` instead.

Reply:
if we're not in debug mode, this assertion won't even run, right?

Reply:
Do we really want a throw StateError here instead of a no-op?
I don't think it would be uncommon for someone to register the callback without a `kDebugMode` guard and then it would their app would crash in production because of the throw while registering.

I could remove the assert and leave the early return. What do you think?